xieyu_wang/Logger.hpp

```diff
@@ -0,0 +1,28 @@
+#pragma once
+#include <fstream>
+#include <mutex>
+#include <optional>
+#include <string>
+
+namespace uni_cpp_practice {
+class Logger {
+ public:
+  static Logger& getLogger() {
+    static Logger logger;
+    return logger;
+  }
+  void setFile(const std::optional<std::string>& filename);
+  void log(const std::string& string);
+  ~Logger();
+
+ private:
+  std::optional<std::ofstream> file_;
+  std::mutex mutex_;
+
+  Logger() = default;
```

Comment:
Again, did you read the task?
You need to add `config.hpp`, and take path to log file from there.
```suggestion
  Logger() : file_(config::LOGGER_FILE_PATH) {
    if (!file_.is_open()) {
      throw std::runtime_error("Failed to open log file");
    }
  }
```
If you do like this, you don't need any `optional` and you don't need `setFile`